### PR TITLE
Feat : pt: support property fitting

### PR DESCRIPTION
**Pull request by Chengqian-Zhang**
Support property fitting:

One can fit multiple scalar properties at the same time (application scenario: predicting homo, lumo, bandgap of a structure at the same time), the number of properties to be fitted is determined by the "task_num" keyword in "model/fitting". Of course, it is possible to fit only 1 scalar property (same as energy).

But at the moment, it only supports *averaging* atomic_property to get the final property, and it doesn't add bias to each element type (different from energy), more comprehensive functions will be realized in the future pr.

You'll notice that I've also modified some parts of denoise, but due to the complexity of denoise fitting and the unfinished refactoring of the loss part @iProzd , the denoise functionality has not been fully implemented in this pr.

### CODE REVIEW

File: deepmd/utils/argcheck.py [hunk 1]
```diff
@@ -1176,6 +1176,82 @@ def fitting_dipole():
...
 
+@fitting_args_plugin.register("property")
+def fitting_property():
+    doc_neuron = "The number of neurons in each hidden layers of the fitting net. When two hidden layers are of the same size, a skip connection is built."
+    doc_activation_function = f'The activation function in the fitting net. Supported activation functions are {list_to_doc(ACTIVATION_FN_DICT.keys())} Note that "gelu" denotes the custom operator version, and "gelu_tf" denotes the TF standard version. If you set "None" or "none" here, no activation function will be used.'
+    doc_resnet_dt = 'Whether to use a "Timestep" in the skip connection'
+    doc_precision = f"The precision of the fitting net parameters, supported options are {list_to_doc(PRECISION_DICT.keys())} Default follows the interface precision."
+    doc_sel_type = "The atom types for which the atomic dipole will be provided. If not set, all types will be selected."
+    doc_seed = "Random seed for parameter initialization of the fitting net"
+    doc_task_num = "The number of outputs of fitting net"
```
Comment (wanghan-iapcm) [inline deepmd/utils/argcheck.py:1187]: the "dimension" instead of "number"  seems to be better

File: deepmd/utils/argcheck.py [hunk 2]
```diff
@@ -1176,6 +1176,82 @@ def fitting_dipole():
     ]
 
 
+@fitting_args_plugin.register("property")
+def fitting_property():
+    doc_neuron = "The number of neurons in each hidden layers of the fitting net. When two hidden layers are of the same size, a skip connection is built."
+    doc_activation_function = f'The activation function in the fitting net. Supported activation functions are {list_to_doc(ACTIVATION_FN_DICT.keys())} Note that "gelu" denotes the custom operator version, and "gelu_tf" denotes the TF standard version. If you set "None" or "none" here, no activation function will be used.'
+    doc_resnet_dt = 'Whether to use a "Timestep" in the skip connection'
+    doc_precision = f"The precision of the fitting net parameters, supported options are {list_to_doc(PRECISION_DICT.keys())} Default follows the interface precision."
+    doc_sel_type = "The atom types for which the atomic dipole will be provided. If not set, all types will be selected."
```
Comment (wanghan-iapcm) [inline deepmd/utils/argcheck.py:1185]: do you really support `sel_type` in property fitting?

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
It is highly recommended to only provide property fitting imple. in this PR and provide all denoise supports in a separate PR.

File: deepmd/pt/model/task/property.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        task_num: int = 1,
+        neuron: List[int] = [128, 128, 128],
+        bias_atom_e: Optional[torch.Tensor] = None,
+        resnet_dt: bool = True,
+        numb_fparam: int = 0,
+        numb_aparam: int = 0,
+        activation_function: str = "tanh",
+        precision: str = DEFAULT_PRECISION,
+        mixed_types: bool = True,
+        **kwargs,
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/property.py:58]: what are expected for kwargs?

**Comment by Chengqian-Zhang** [comment]
@wanghan-iapcm  Seemly I do not have the necessary permissions to add reviewers to a pull request in the repository. Can you give me higher access?
Rjiexf3OdW

File: deepmd/infer/deep_property.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        cells: Optional[np.ndarray],
+        atom_types: Union[List[int], np.ndarray],
+        atomic: bool = False,
+        fparam: Optional[np.ndarray] = None,
+        aparam: Optional[np.ndarray] = None,
+        mixed_type: bool = False,
+        **kwargs: Dict[str, Any],
+    ) -> Tuple[np.ndarray, ...]:
+        """Evaluate energy, force, and virial. If atomic is True,
+        also return atomic energy and atomic virial.
```
Comment (njzjz) [inline deepmd/infer/deep_property.py:76]: The doc here seems wrong.

File: deepmd/infer/deep_property.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        coords : np.ndarray
+            The coordinates of the atoms, in shape (nframes, natoms, 3).
+        cells : np.ndarray
+            The cell vectors of the system, in shape (nframes, 9). If the system
+            is not periodic, set it to None.
+        atom_types : List[int] or np.ndarray
+            The types of the atoms. If mixed_type is False, the shape is (natoms,);
+            otherwise, the shape is (nframes, natoms).
+        atomic : bool, optional
+            Whether to return atomic energy and atomic virial, by default False.
```
Comment (njzjz) [inline deepmd/infer/deep_property.py:89]: The doc here seems wrong.

File: deepmd/pt/loss/property.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        """Return loss on loss and force.
+
+        Args:
+        - model_pred: Property prediction.
+        - label: Target property.
+        - natoms: Tell atom count.
+
+        Returns
+        -------
+        - loss: Loss to minimize.
```
Comment (njzjz) [inline deepmd/pt/loss/property.py:62]: Should use the same format as docs for other methods.

File: deepmd/pt/model/model/property_model.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                model_predict["extended_force"] = model_ret["energy_derv_r"].squeeze(-2)
+            if self.do_grad_c("energy"):
+                model_predict["virial"] = model_ret["energy_derv_c_redu"].squeeze(-2)
+                if do_atomic_virial:
+                    model_predict["extended_virial"] = model_ret[
+                        "energy_derv_c"
+                    ].squeeze(-3)
+            else:
+                assert model_ret["dforce"] is not None
+                model_predict["dforce"] = model_ret["dforce"]
```
Comment (njzjz) [inline deepmd/pt/model/model/property_model.py:82]: The keys here seem different from those in `forward`...

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/utils/argcheck.py (hunk 2 above)
Comment (Chengqian-Zhang) [reply deepmd/utils/argcheck.py]: I didn't support "sel_type". I have deleted it.

File: deepmd/pt/model/task/property.py (hunk 3 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/model/task/property.py]: I've seen other tasks(ener.py) written this way so I've written it the same way. But I realized it's not really useful here now, so I deleted it.

File: deepmd/infer/deep_property.py (hunk 4 above)
Comment (Chengqian-Zhang) [reply deepmd/infer/deep_property.py]: I have modified it.

File: deepmd/infer/deep_property.py (hunk 5 above)
Comment (Chengqian-Zhang) [reply deepmd/infer/deep_property.py]: I have modified it.

File: deepmd/pt/loss/property.py (hunk 6 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/loss/property.py]: I have modified it.

File: deepmd/pt/model/model/property_model.py (hunk 7 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/model/model/property_model.py]: I have modified it.

File: deepmd/pt/model/task/property.py (hunk 3 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/model/task/property.py:58]: like "seed" used to init its father class "InvarFitting"

**Comment by Chengqian-Zhang** [comment]
@wanghan-iapcm  I have delete all denoise content.

**Comment by iProzd** [comment]
> Support property fitting:
> 
> One can fit multiple scalar properties at the same time (application scenario: predicting homo, lumo, bandgap of a structure at the same time), the number of properties to be fitted is determined by the "task_num" keyword in "model/fitting". Of course, it is possible to fit only 1 scalar property (same as energy).
> 
> But at the moment, it only supports _averaging_ atomic_property to get the final property, and it doesn't add bias to each element type (different from energy), more comprehensive functions will be realized in the future pr.
> 
> You'll notice that I've also modified some parts of denoise, but due to the complexity of denoise fitting and the unfinished refactoring of the loss part @iProzd , the denoise functionality has not been fully implemented in this pr.

You can refactor denoise loss when https://github.com/deepmodeling/deepmd-kit/pull/3569 is merged.

File: deepmd/utils/argcheck.py [hunk 8]
```diff
@@ -1176,6 +1176,37 @@ def fitting_dipole():
...
 
 
+@fitting_args_plugin.register("property")
+def fitting_property():
+    doc_neuron = "The number of neurons in each hidden layers of the fitting net. When two hidden layers are of the same size, a skip connection is built."
+    doc_activation_function = f'The activation function in the fitting net. Supported activation functions are {list_to_doc(ACTIVATION_FN_DICT.keys())} Note that "gelu" denotes the custom operator version, and "gelu_tf" denotes the TF standard version. If you set "None" or "none" here, no activation function will be used.'
+    doc_resnet_dt = 'Whether to use a "Timestep" in the skip connection'
+    doc_precision = f"The precision of the fitting net parameters, supported options are {list_to_doc(PRECISION_DICT.keys())} Default follows the interface precision."
+    doc_seed = "Random seed for parameter initialization of the fitting net"
+    doc_task_num = "The dimension of outputs of fitting net"
```
Comment (wanghan-iapcm) [inline deepmd/utils/argcheck.py:1186]: I mean the variable name should be self-explanary like "task_dim" instead of "task_num"

File: deepmd/dpmodel/fitting/property_fitting.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    def deserialize(cls, data: dict) -> "GeneralFitting":
+        data = copy.deepcopy(data)
+        check_version_compatibility(data.pop("@version", 1), 1, 1)
+        data.pop("var_name")
+        data.pop("dim_out")
+        return super().deserialize(data)
+
+    def serialize(self) -> dict:
+        """Serialize the fitting to dict."""
+        return {**super().serialize(), "type": "property", "task_num": self.task_num}
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/fitting/property_fitting.py:94]: no need to store task_dim in serialization. the information is already serialized by invar_fitting.

File: deepmd/pt/loss/property.py
Comment (wanghan-iapcm) [inline deepmd/pt/loss/property.py:1]: better to commit loss after merging #3569 by a new PR.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/dpmodel/fitting/property_fitting.py (hunk 9 above)
Comment (Chengqian-Zhang) [reply deepmd/dpmodel/fitting/property_fitting.py]: It is necessary because the key `task_dim` is needed to initialize `PropertyFittingNet` when deserializing.

File: deepmd/pt/loss/property.py
Comment (Chengqian-Zhang) [reply deepmd/pt/loss/property.py:1]: I have deleted loss part

File: deepmd/pt/model/task/property.py (hunk 3 above)
Comment (wanghan-iapcm) [reply deepmd/pt/model/task/property.py:58]: you need to explicitly write all the arguments. 
plz add doc

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: examples/property/train/input_torch.json
Comment (njzjz) [inline examples/property/train/input_torch.json:1]: Please add this file to `source/tests/common/test_examples.py`.

File: deepmd/utils/argcheck.py [hunk 10]
```diff
@@ -1176,6 +1176,37 @@ def fitting_dipole():
     ]
 
 
+@fitting_args_plugin.register("property")
```
Comment (njzjz) [inline deepmd/utils/argcheck.py:1179]: Need to mark as pytorch only

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/utils/argcheck.py (hunk 10 above)
Comment (njzjz) [reply deepmd/utils/argcheck.py:1179]: see line 616

**Comment by Chengqian-Zhang** [comment]
I think dptest also needs to be implemented in this PR?

**Comment by wanghan-iapcm** [comment]
the output stat should be replace by the `change_out_bias` method provided in #3480

**Comment by wanghan-iapcm** [comment]
#3480 is merging @Chengqian-Zhang please update the pr accordingly.

**Comment by Chengqian-Zhang** [comment]
@wanghan-iapcm  Copy that, I will update immediately.

**Comment by Chengqian-Zhang** [comment]
I will complete this function in another PR #3867
